File: android/ffmpeg-kit-next-android-lib/src/main/kotlin/com/arthenica/ffmpegkit/FFprobeSession.kt
/*
 * Copyright (c) 2020-2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

package com.arthenica.ffmpegkit

/**
 * <p>An FFprobe session.
 *
 * @param arguments command arguments as an array
 * @param completeCallback specific complete callback
 * @param logCallback session specific log callback
 * @param logRedirectionStrategy session specific log redirection strategy
 */
open class FFprobeSession private constructor(
    arguments: Array<String>,
    private val completeCallback: FFprobeSessionCompleteCallback?,
    logCallback: LogCallback?,
    logRedirectionStrategy: LogRedirectionStrategy
) : AbstractSession(arguments, logCallback, logRedirectionStrategy), Session {

    /**
     * Returns the session specific complete callback.
     *
     * @return session specific complete callback
     */
    open fun getCompleteCallback(): FFprobeSessionCompleteCallback? = completeCallback

    override fun isFFmpeg(): Boolean = false

    override fun isFFprobe(): Boolean = true

    override fun isMediaInformation(): Boolean = false

    override fun toString(): String = buildString {
        append("FFprobeSession{")
        append("sessionId=")
        append(sessionId)
        append(", createTime=")
        append(createTime)
        append(", startTime=")
        append(startTime)
        append(", endTime=")
        append(endTime)
        append(", arguments=")
        append(FFmpegKitConfig.argumentsToString(arguments))
        append(", logs=")
        append(getLogsAsString())
        append(", state=")
        append(state)
        append(", returnCode=")
        append(returnCode)
        append(", failStackTrace=")
        append('\'')
        append(failStackTrace)
        append('\'')
        append('}')
    }

    companion object {

        /**
         * Builds a new FFprobe session.
         *
         * @param arguments command arguments
         * @return created session
         */
        @JvmStatic
        fun create(arguments: Array<String>): FFprobeSession =
            FFprobeSession(arguments, null, null, FFmpegKitConfig.getLogRedirectionStrategy())

        /**
         * Builds a new FFprobe session.
         *
         * @param arguments        command arguments
         * @param completeCallback session specific complete callback
         * @return created session
         */
        @JvmStatic
        fun create(
            arguments: Array<String>,
            completeCallback: FFprobeSessionCompleteCallback?
        ): FFprobeSession =
            FFprobeSession(
                arguments,
                completeCallback,
                null,
                FFmpegKitConfig.getLogRedirectionStrategy()
            )

        /**
         * Builds a new FFprobe session.
         *
         * @param arguments        command arguments
         * @param completeCallback session specific complete callback
         * @param logCallback      session specific log callback
         * @return created session
         */
        @JvmStatic
        fun create(
            arguments: Array<String>,
            completeCallback: FFprobeSessionCompleteCallback?,
            logCallback: LogCallback?
        ): FFprobeSession =
            FFprobeSession(
                arguments,
                completeCallback,
                logCallback,
                FFmpegKitConfig.getLogRedirectionStrategy()
            )

        /**
         * Builds a new FFprobe session.
         *
         * @param arguments              command arguments
         * @param completeCallback       session specific complete callback
         * @param logCallback            session specific log callback
         * @param logRedirectionStrategy session specific log redirection strategy
         * @return created session
         */
        @JvmStatic
        fun create(
            arguments: Array<String>,
            completeCallback: FFprobeSessionCompleteCallback?,
            logCallback: LogCallback?,
            logRedirectionStrategy: LogRedirectionStrategy
        ): FFprobeSession =
            FFprobeSession(arguments, completeCallback, logCallback, logRedirectionStrategy)
    }
}

File: windows/src/internal/FFprobeSession.cpp
/*
 * Copyright (c) 2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "FFprobeSession.h"
#include "FFmpegKitConfig.h"
#include "LogCallback.h"

extern void
addSessionToSessionHistory(const std::shared_ptr<ffmpegkit::internal::Session> session);

std::shared_ptr<ffmpegkit::internal::FFprobeSession>
ffmpegkit::internal::FFprobeSession::create(const std::list<std::string> &arguments) {
    auto session = std::static_pointer_cast<ffmpegkit::internal::FFprobeSession>(
        std::make_shared<ffmpegkit::internal::FFprobeSession::PublicFFprobeSession>(
            arguments, nullptr, nullptr,
            ffmpegkit::internal::FFmpegKitConfig::getLogRedirectionStrategy()));
    addSessionToSessionHistory(session);
    return session;
}

std::shared_ptr<ffmpegkit::internal::FFprobeSession> ffmpegkit::internal::FFprobeSession::create(
    const std::list<std::string> &arguments,
    const FFprobeSessionCompleteCallback completeCallback) {
    auto session = std::static_pointer_cast<ffmpegkit::internal::FFprobeSession>(
        std::make_shared<ffmpegkit::internal::FFprobeSession::PublicFFprobeSession>(
            arguments, completeCallback, nullptr,
            ffmpegkit::internal::FFmpegKitConfig::getLogRedirectionStrategy()));
    addSessionToSessionHistory(session);
    return session;
}

std::shared_ptr<ffmpegkit::internal::FFprobeSession> ffmpegkit::internal::FFprobeSession::create(
    const std::list<std::string> &arguments,
    const FFprobeSessionCompleteCallback completeCallback,
    const ffmpegkit::internal::LogCallback logCallback) {
    auto session = std::static_pointer_cast<ffmpegkit::internal::FFprobeSession>(
        std::make_shared<ffmpegkit::internal::FFprobeSession::PublicFFprobeSession>(
            arguments, completeCallback, logCallback,
            ffmpegkit::internal::FFmpegKitConfig::getLogRedirectionStrategy()));
    addSessionToSessionHistory(session);
    return session;
}

std::shared_ptr<ffmpegkit::internal::FFprobeSession> ffmpegkit::internal::FFprobeSession::create(
    const std::list<std::string> &arguments,
    const FFprobeSessionCompleteCallback completeCallback,
    const ffmpegkit::internal::LogCallback logCallback,
    const LogRedirectionStrategy logRedirectionStrategy) {
    auto session = std::static_pointer_cast<ffmpegkit::internal::FFprobeSession>(
        std::make_shared<ffmpegkit::internal::FFprobeSession::PublicFFprobeSession>(
            arguments, completeCallback, logCallback, logRedirectionStrategy));
    addSessionToSessionHistory(session);
    return session;
}

struct ffmpegkit::internal::FFprobeSession::PublicFFprobeSession
    : public ffmpegkit::internal::FFprobeSession {
    PublicFFprobeSession(const std::list<std::string> &arguments,
                         const FFprobeSessionCompleteCallback completeCallback,
                         const ffmpegkit::internal::LogCallback logCallback,
                         const LogRedirectionStrategy logRedirectionStrategy)
        : FFprobeSession(arguments, completeCallback, logCallback,
                         logRedirectionStrategy) {}
};

ffmpegkit::internal::FFprobeSession::FFprobeSession(
    const std::list<std::string> &arguments,
    const FFprobeSessionCompleteCallback completeCallback,
    const ffmpegkit::internal::LogCallback logCallback,
    const LogRedirectionStrategy logRedirectionStrategy)
    : ffmpegkit::internal::AbstractSession(arguments, logCallback,
                                 logRedirectionStrategy),
      _completeCallback{completeCallback} {}

ffmpegkit::internal::FFprobeSessionCompleteCallback
ffmpegkit::internal::FFprobeSession::getCompleteCallback() {
    return _completeCallback;
}

bool ffmpegkit::internal::FFprobeSession::isFFmpeg() const { return false; }

bool ffmpegkit::internal::FFprobeSession::isFFprobe() const { return true; }

bool ffmpegkit::internal::FFprobeSession::isMediaInformation() const { return false; }
